app/controllers/recipes_controller.rb

```diff
@@ -0,0 +1,73 @@
+class RecipesController < ApplicationController
+  before_action :authenticate_user!
+
+  def index
+    @user = current_user
+    @recipes = @user.recipes.includes(:user).all.paginate(page: params[:page], per_page: 3) # preload all users
+  end
+
+  def new
+    @user = current_user
+    @recipe = @user.recipes.build
+  end
+
+  def create
+    @user = current_user
+    @recipe = @user.recipes.build(recipe_params)
+
+    if @recipe.save
+      redirect_to root_path(@user), notice: 'Recipe was successfully created!'
+    else
+      render :new
+    end
+  end
+
+  def destroy
+    @user = User.find(params[:user_id])
+    @recipe = Recipe.find(params[:id])
+    if @recipe.destroy
+      redirect_to user_recipes_path(@recipe.user), notice: 'Recipe was successfully deleted.'
+    else
+      redirect_to user_recipes_path(current_user), alert: 'Recipe could not be destroyed.'
+    end
+  end
+
+  def public_recipes
+    @public_recipes = Recipe.includes(:user).where(public: true).order(created_at: :desc)
+      .paginate(page: params[:page], per_page: 5)
+  end
+
+  def show
+    @user = current_user
+    @recipe = Recipe.includes(recipe_foods: :food).find(params[:id]) # Preload recipe_foods
+    @recipefoods = @recipe.recipe_foods
+  end
+
+  def generate_shopping_list
+    @recipe = Recipe.find(params[:id])
+    @missing_food_items = find_missing_food_items(@recipe)
+    @total_food_to_buy = @missing_food_items.count
+    @total_value_of_food_needed = @missing_food_items.sum { |item| item[:quantity] * item[:price] }
+    @missing_food_items = Food.where(id: @missing_food_items.map(&:id)).includes(:user) # Preload users
+  end
```

Comment:
- [ ] Please kindly ensure that generate_shopping_list is also part of the navbar items as state in the requirement. 🙏 

<img width="481" alt="Untitled" src="https://github.com/jcmunav63/rails-italian-recipes-app/assets/103386183/cd909cd4-b1ff-4bf7-a20b-92d1be40e750">